### ISSUE TITLE
Tapped triggered after Drag

### ISSUE THREAD

**Issue by stuartmscott** [fyne-io/fyne#2235]
### Describe the bug:
Tapped is triggered after Drag which can interfere with the user interface; consider the selection scenario where the user can tap to select, or drag to define a rectangle selection. The user drags out a selection but after they release the mouse the selection gets dropped and instead the single element under the mouse gets selected.

### Example code:
```go
package main

import (
	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/app"
	"fyne.io/fyne/v2/canvas"
	"fyne.io/fyne/v2/theme"
	"fyne.io/fyne/v2/widget"
	"image/color"
	"log"
)

func main() {
	a := app.New()
	w := a.NewWindow("Events")
	w.SetContent(&interactiveRect{})
	w.Resize(fyne.NewSize(200, 200))
	w.ShowAndRun()
}

var _ fyne.Tappable = (*interactiveRect)(nil)
var _ fyne.SecondaryTappable = (*interactiveRect)(nil)
var _ fyne.Draggable = (*interactiveRect)(nil)

type interactiveRect struct {
	widget.BaseWidget
}

func (r *interactiveRect) CreateRenderer() fyne.WidgetRenderer {
	r.ExtendBaseWidget(r)
	return &interactiveRectRenderer{
		rect: canvas.NewRectangle(theme.PrimaryColor()),
	}
}

func (r *interactiveRect) MinSize() fyne.Size {
	r.ExtendBaseWidget(r)
	return r.BaseWidget.MinSize()
}

func (r *interactiveRect) Tapped(event *fyne.PointEvent) {
	log.Println("Tapped", event)
}

func (r *interactiveRect) TappedSecondary(event *fyne.PointEvent) {
	log.Println("TappedSecondary", event)
}

func (r *interactiveRect) Dragged(event *fyne.DragEvent) {
	log.Println("Dragged", event)
}

func (r *interactiveRect) DragEnd() {
	log.Println("DragEnd")
}

var _ fyne.WidgetRenderer = (*interactiveRectRenderer)(nil)

type interactiveRectRenderer struct {
	rect *canvas.Rectangle
}

func (r *interactiveRectRenderer) BackgroundColor() color.Color {
	return color.Transparent
}

func (r *interactiveRectRenderer) Destroy() {
}

func (r *interactiveRectRenderer) Layout(size fyne.Size) {
	r.rect.Resize(size)
}

func (r *interactiveRectRenderer) MinSize() fyne.Size {
	return r.rect.MinSize()
}

func (r *interactiveRectRenderer) Objects() []fyne.CanvasObject {
	return []fyne.CanvasObject{r.rect}
}

func (r *interactiveRectRenderer) Refresh() {
	r.rect.Refresh()
}
```

### Aside:
In v1.3.3 the Tapped is triggered after the last Drag, but before the DragEnd so it was possible to maintain a `dragging bool`  and ignore Tapped if dragging == true.

v2.0.3
```
2021/05/10 14:15:01 Dragged &{{{130 75} {126 71}} {-1 0}}
2021/05/10 14:15:01 Dragged &{{{125 75} {121 71}} {-5 0}}
2021/05/10 14:15:01 Dragged &{{{119 75} {115 71}} {-6 0}}
2021/05/10 14:15:01 Dragged &{{{110 76} {106 72}} {-9 1}}
2021/05/10 14:15:01 Dragged &{{{106 77} {102 73}} {-4 1}}
2021/05/10 14:15:01 Dragged &{{{101 78} {97 74}} {-5 1}}
2021/05/10 14:15:01 Dragged &{{{100 79} {96 75}} {-1 1}}
2021/05/10 14:15:01 Dragged &{{{98 80} {94 76}} {-2 1}}
2021/05/10 14:15:01 Dragged &{{{97 80} {93 76}} {-1 0}}
2021/05/10 14:15:01 Dragged &{{{97 80} {93 76}} {0 0}}
2021/05/10 14:15:01 Dragged &{{{97 81} {93 77}} {0 1}}
2021/05/10 14:15:01 Dragged &{{{96 81} {92 77}} {-1 0}}
2021/05/10 14:15:01 DragEnd
2021/05/10 14:15:01 Tapped &{{96 81} {92 77}}
```

v1.3.3
```
2021/05/10 14:20:42 Dragged &{{{0 0} {88 45}} -1 0}
2021/05/10 14:20:42 Dragged &{{{0 0} {86 48}} -2 3}
2021/05/10 14:20:42 Dragged &{{{0 0} {85 50}} -1 2}
2021/05/10 14:20:42 Dragged &{{{0 0} {83 52}} -2 2}
2021/05/10 14:20:42 Dragged &{{{0 0} {81 54}} -2 2}
2021/05/10 14:20:42 Dragged &{{{0 0} {78 58}} -3 4}
2021/05/10 14:20:42 Dragged &{{{0 0} {77 59}} -1 1}
2021/05/10 14:20:42 Dragged &{{{0 0} {76 61}} -1 2}
2021/05/10 14:20:42 Dragged &{{{0 0} {74 64}} -2 3}
2021/05/10 14:20:42 Dragged &{{{0 0} {71 67}} -3 3}
2021/05/10 14:20:42 Dragged &{{{0 0} {70 70}} -1 3}
2021/05/10 14:20:42 Dragged &{{{0 0} {68 74}} -2 4}
2021/05/10 14:20:42 Dragged &{{{0 0} {67 75}} -1 1}
2021/05/10 14:20:42 Dragged &{{{0 0} {67 75}} 0 0}
2021/05/10 14:20:42 Tapped &{{71 79} {67 75}}
2021/05/10 14:20:42 DragEnd
```

### Device (please complete the following information):
 - **OS:** MacOS
 - **Version:** Big Sur
 - **Go version:** 1.15
 - **Fyne version:** 76e0133

**Comment by stuartmscott** [comment]
> This is really strange. I cannot replicate it within the Entry widget adding debug lines on release/v2.0.x or develop branches.

Can you reproduce with the provided code sample?

I can reproduce with a go.mod containing `require fyne.io/fyne/v2 v2.0.3` and `replace fyne.io/fyne/v2 => ../../../fyne.io/fyne` where the replacement points to a repo holding develop (76e0133)

> P.S. your example is not using the version of Fyne you think it is - that's v1

It had the v2 suffix when I ran it with v2.0.3, but then I removed it and changed the go.mod so I could test with v1.3.3.

**Comment by andydotxyz** [comment]
> Can you reproduce with the provided code sample?

yes I can. But I can’t quite understand how it happens here but not in entry. More work required to hunt down the cause I think.

> but then I removed it and changed the go.mod so I could test with v1.3.3.

goodness I’m sorry, misread your message - I was convinced you were reporting against develop and v2.0.3. Oops.

**Comment by stuartmscott** [comment]
I wonder if it is because Entry implements more interfaces than this example (ie. the desktop MouseIn/Move/Out) so it takes a slightly different path through the driver

**Comment by andydotxyz** [comment]
You are probably right about that. I have tagged 2.0.x in case we get a simple fix on this before any other patch releases go out.

**Comment by s77rt** [comment]
not sure i follow correctly but why not just:
https://github.com/fyne-io/fyne/blob/3814a66f75ff5d24be6fa107f1dfb9c4faa859b1/internal/driver/glfw/window.go#L823
to:
`				if button == desktop.MouseButtonSecondary && altTap && !mouseDragStarted {`

and
https://github.com/fyne-io/fyne/blob/3814a66f75ff5d24be6fa107f1dfb9c4faa859b1/internal/driver/glfw/window.go#L831
to:
`	if action == glfw.Release && button == desktop.MouseButtonPrimary && !mouseDragStarted {`

**Comment by andydotxyz** [comment]
I think you are probably right @s77rt

**Comment by andydotxyz** [comment]
On `develop` for testing.

**Comment by andydotxyz** [comment]
Merged to `release/v2.0.x` as well
